keps/sig-instrumentation/20190605-metrics-validation-and-verification.md

```diff
@@ -0,0 +1,180 @@
+Stable metrics will use exactly the same functions as normal once, but code defining them will need to comply to specific format:
+
+* Metric options (`CounterOpts`) needs to be directly passed to function (`kubemetrics.NewCounterVec`).
+* Metric arguments can only be set to values. Fields cannot be set to const, variables or function result (apart of `kubemetrics.STABLE`). 
+```go
+var someCounterVecMetric = kubemetrics.NewCounterVec(
+  &prometheus.CounterOpts{
+    Name:           "some_metric",
+    Help:           "some description",
+    StabilityLevel: kubemetrics.STABLE,
+  },
+  []string{"some-label", "other-label"},
+}
+```
+Package name `kubemetrics` can be replaced with local name of framework import.
+
+Those restrictions will allow AST based analysis to correctly interpret metric definitions.
+
+### Stable metric detection strategy 
+
+Static analysis will be done in two steps. First find stable metrics, second read and validate their fields. 
+
+In first one step we will want to distinguish stable metric without relying on their definition structure, allowing freedom of declaration for non-stable metrics. This will be done by finding occurrences of metric options object initialization (`CounterOpts` for `Counter`). If this initialization sets `StabilityLevel` then it will validate and read it. If value could not be extracted then script should fail. Metrics which set `kubemetrics.STABLE` will be passed to second step.
+
+Second step will be extracting information from stable metric definitions. This step will be expecting exact structure of AST subtree of new metric call matching format proposed above. If metric options object was initialized outside of new metric call or call structure deviates from expected format then static analysis should fail. This restriction will ensure that no field was read incorrectly or stable metric missed.
+
+
+### Format of stable metrics list file
+
+This file should include information about metrics needed to detect a breaking change. Those fields should include `name`, `namespace`, `subsystem` as they are used to build metric fully-qualified name, metric type and metric labels. Metric labels should include both vector labels and const labels. 
+
+Metric information stored in file should be in easly readable and reviewable format. For that pretty printed json will be used. Example:
+```json
+[
+  {
+    "name": "some_metric",
+    "subsystem": "some_subsystem",
+    "namespace": "some_namespace",
+    "labels": [
+      "some-label"
+    ],
+    "type": "Counter"
+  }
+]
+```
+
+Json keys and lists will be sorted alphabetically to ensure that there is only one correct output.
+
+### Failure Modes
+
+In cases where static analysis could not be done correctly (invalid stable metric format, etc.) script should fail (non zero response code) and write to STDERR information on problem. Example:
```

Comment:
In cases where static analysis could not be done correctly [on a stable metric]

Reply:
There is small restrictions for non-stable metrics,  `StabilityLevel` cannot be auto-generated as we need it to recognize stable ones.